eda/src/autogluon/eda/analysis/shift.py

```diff
@@ -0,0 +1,347 @@
+            - a tuple of source dataframe and target dataframe
+        """
+        if isinstance(data, pd.DataFrame):
+            sample_label = self.sample_label
+            assert sample_label in data.columns, "sample_label needs to be a column of data"
+            assert self.split, "sample_label requires the split parameter"
+            data = data.copy() # makes a copy
+        else:
+            assert len(data) == 2, "Data needs to be tuple/list of (source, target) if sample_label is None"
+            data = self._make_source_target_label(data, self.sample_label) # makes a copy
+        if data.index.has_duplicates:
+            data.index = pd.RangeIndex(data.shape[0])
+        train = data.sample(frac=self.split)
+        test = data.drop(train.index)
+        self.classifier.fit(train, **kwargs)
+        yhat = self.classifier.predict(test)
+        self.test_stat = self.eval_metric(test[self.sample_label], yhat)
+        self.has_fi = (getattr(self.classifier, "feature_importance", None) is not None)
+        if self.has_fi and self.compute_fi:
+            self._test = test  # for feature importance
+        self._is_fit = True
+
+    @post_fit
+    def _pvalue_half_permutation(self,
+                                 num_permutations=1000):
+        """The half permutation method for computing p-values.
+        See Section 9.1 of https://arxiv.org/pdf/1602.02210.pdf
+        """
+        perm_stats = [self.test_stat]
+        yhat = self.classifier.predict(self._test)
+        for i in range(num_permutations):
+            perm_yhat = np.random.permutation(yhat)
+            perm_test_stat = self.eval_metric(
+                self._test[self.sample_label],
+                perm_yhat
+            )
+            perm_stats.append(perm_test_stat)
+        pval = (self.test_stat <= np.array(perm_stats)).mean()
+        return pval
+
+    @post_fit
+    def pvalue(self,
+               num_permutations: int=1000):
+        """Compute the p-value which measures the significance level for the test statistic
+
+        Parameters
+        ----------
+        num_permutations: int, default = 1000
+            The number of permutations used for any permutation based method
+
+        Returns
+        -------
+        float of the p-value for the 2-sample test
+        """
+        pval = self._pvalue_half_permutation(num_permutations=num_permutations)
+        return pval
+
+    @post_fit
+    def feature_importance(self):
+        """Returns the feature importances for the trained classifier for source v. target
+
+        Returns
+        -------
+        pd.DataFrame of feature importances
+        """
+        assert self.has_fi, "Classifier class does not have feature_importance method"
+        assert self.compute_fi, "Set compute_fi to True to compute feature importances"
+        fi_scores = self.classifier.feature_importance(self._test)
+        return fi_scores
+
+
+class C2STShiftDetector:
```

Comment:
This class is redundant: data preprocessing logic can be inside `Classifier2ST` class while `decision()/results()` can be in Analysis/Visualization classes.

Reply:
yeah, this is also a holdout from before - if you think that it is important I can do this

Reply:
it is additional layer to maintain if we can remove it, we should